druid/src/widget/focus.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 The druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A focus widget.
+
+use druid::widget::prelude::*;
+
+use druid::{
+    commands, Command, Data, FocusNode, HotKey, KbKey, Point, Rect, SysMods, Widget, WidgetPod,
+};
+
+/// A widget that manages a FocusNode to allow focus to be given to this widget and its descendants.
+pub struct Focus<T> {
+    child: WidgetPod<T, Box<dyn Widget<T>>>,
+    auto_focus: bool,
+    focus_requested: bool,
+    focus_node: FocusNode,
+}
+
+impl<T: Data> Focus<T> {
+    /// Create a new Focus widget with a child
+    pub fn new(child: impl Widget<T> + 'static) -> Self {
+        Focus {
+            child: WidgetPod::new(child).boxed(),
+            focus_node: FocusNode::empty(),
+            auto_focus: false,
+            focus_requested: false,
+        }
+    }
+
+    /// Builder-style method to set the `Focus`'s auto focus.
+    pub fn with_auto_focus(mut self, auto_focus: bool) -> Self {
+        self.auto_focus = auto_focus;
+        self
+    }
+}
+
+impl<T: Data> Widget<T> for Focus<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        let previous_focus_node = ctx.focus_node();
+        ctx.set_focus_node(self.focus_node);
+
+        if self.auto_focus && !self.focus_requested {
+            self.focus_requested = true;
+            ctx.request_focus();
+        }
+
+        match event {
+            Event::MouseDown(_) => {
+                ctx.request_focus();
+                ctx.request_paint();
+            }
+            Event::KeyDown(key_event) => {
+                match key_event {
+                    // Tab and shift+tab
+                    k_e if HotKey::new(None, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_next();
+                    }
+                    k_e if HotKey::new(SysMods::Shift, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_prev();
+                    }
+                    _ => (),
```

Comment:
I think we should send the key event to our child first, and then only handle it ourselves if `is_handled() == false`; this would let a text field decide whether or not it wanted to insert a tab character or not.

Reply:
Ok